x-pack/plugins/security_solution/server/endpoint/services/artifacts/manifest_manager/manifest_manager.ts

```diff
@@ -515,4 +515,68 @@ export class ManifestManager {
   public getArtifactsClient(): EndpointArtifactClientInterface {
     return this.artifactClient;
   }
+
+  /**
+   * Cleanup .fleet-artifacts index if there are some orphan artifacts
+   */
+  public async cleanup(manifest: Manifest) {
+    try {
+      const fleetArtifacts = [];
+      const perPage = 100;
+      let page = 1;
+
+      let fleetArtifactsResponse = await this.artifactClient.listArtifacts({
+        perPage,
+        page,
+      });
+      fleetArtifacts.push(...fleetArtifactsResponse.items);
+
+      while (
+        fleetArtifactsResponse.total > fleetArtifacts.length &&
+        !isEmpty(fleetArtifactsResponse.items)
+      ) {
+        page += 1;
+        fleetArtifactsResponse = await this.artifactClient.listArtifacts({
+          perPage,
+          page,
+        });
+        fleetArtifacts.push(...fleetArtifactsResponse.items);
+      }
+
+      if (isEmpty(fleetArtifacts)) {
+        return;
+      }
+
+      const artifactsToDelete = [];
+      const badArtifacts = [];
+
+      const manifestArtifactsIds = manifest
+        .getAllArtifacts()
+        .map((artifact) => getArtifactId(artifact));
+
+      for (const fleetArtifact of fleetArtifacts) {
+        const artifactId = getArtifactId(fleetArtifact);
+        const isArtifactInManifest = manifestArtifactsIds.includes(artifactId);
+
+        if (!isArtifactInManifest) {
+          artifactsToDelete.push(this.artifactClient.deleteArtifact(artifactId));
```

Comment:
ℹ️  It should be ok (🤞 ), but I wonder if we should throttle the amount of concurrent request to ES here.

Reply:
Good catch. In the worst case (without artifacts by policy) there shouldn't be more than 9 artifacts to be removed right? Trustad apps, event filters and endpoint exceptions for each OS

Reply:
Right - for now. but we'll be adding more as we go. Like I said, probably should be ok.

Reply:
I can create an issue for 7.16 and take care of concurrent requests once we have artifacts by policy but I think this should be ok for 7.15.1 right?

Reply:
Agreed. Don't hold this up based on this comment.

if you do create a new issue, we should also probably see if we handle the error (log it) for each deletion and allow the `Promise.all()` to fulfill. Right now, the `Promise.all()` will trigger the `catch()` below on the first error it encounters, which means that any subsequent (queued or in flight) delete request would be suppressed to the logs